h2o-persist-hdfs/src/main/java/water/persist/security/HdfsDelegationTokenRefresher.java

```diff
@@ -0,0 +1,230 @@
+package water.persist.security;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.hadoop.security.token.delegation.AbstractDelegationTokenIdentifier;
+import water.H2O;
+import water.MRTask;
+import water.Paxos;
+import water.persist.PersistHdfs;
+import water.util.BinaryFileTransfer;
+import water.util.FileUtils;
+
+import java.io.*;
+import java.security.PrivilegedExceptionAction;
+import java.util.Arrays;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class HdfsDelegationTokenRefresher implements Runnable {
+
+    public static final String H2O_AUTH_USER = "h2o.auth.user";
+    public static final String H2O_AUTH_PRINCIPAL = "h2o.auth.principal";
+    public static final String H2O_AUTH_KEYTAB = "h2o.auth.keytab";
+    public static final String H2O_AUTH_TOKEN_REFRESHER_ENABLED = "h2o.auth.tokenRefresher.enabled";
+    public static final String H2O_AUTH_TOKEN_REFRESHER_INTERVAL_RATIO = "h2o.auth.tokenRefresher.intervalRatio";
+    public static final String H2O_AUTH_TOKEN_REFRESHER_MAX_ATTEMPTS = "h2o.auth.tokenRefresher.maxAttempts";
+    public static final String H2O_AUTH_TOKEN_REFRESHER_RETRY_DELAY_SECS = "h2o.auth.tokenRefresher.retryDelaySecs";
+    public static final String H2O_AUTH_TOKEN_REFRESHER_FALLBACK_INTERVAL_SECS = "h2o.auth.tokenRefresher.fallbackIntervalSecs";
+
+    public static void setup(Configuration conf, String tmpDir) throws IOException {
+        boolean enabled = conf.getBoolean(H2O_AUTH_TOKEN_REFRESHER_ENABLED, false);
+        if (!enabled) {
+            log("HDFS Token renewal is not enabled in configuration", null);
+            return;
+        }
+        String authUser = conf.get(H2O_AUTH_USER);
+        String authPrincipal = conf.get(H2O_AUTH_PRINCIPAL);
+        String authKeytab = conf.get(H2O_AUTH_KEYTAB);
+        if (authKeytab == null) {
```

Comment:
Shouldn't be here also a similar condition for `authPrincipal` and eventually for `authUser`?

Reply:
good point, will fix (authUser can be undefined, in that case we use authPrincipal)

Reply:
fixed, thx